### DISCUSSION TITLE
Help with machines with generic types and typegen

### DISCUSSION THREAD

**Discussion by SimeonC** [statelyai/xstate#3164] [Q&A] [answered]
On my quest to make everything xstate, I've tried utilising generics to build up certain machines, like my "tab form machine" (It's a react form that needs to validate all data and switch to the correct tab to focus the error field). In the case of a form machine I don't want to define the "form structure" up front, as that becomes unusable. Hence I'm trying to use generics to build the context. This worked quite well before using the typegen, but with the typegen I get a bunch of typing mismatches when trying to workaround the generics.

Here's a [TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOASWOyjHVoDEoIQKBlGLXfI0uAL5wAZixBwA5AA8Azl2wSA3AChQkWIjgBXGWmq16aGBR1p2aADZpkMaIJFjJAAVny0AenT4YS1eGjwAFRwGDJwAEqYNg6skl42vsowAJ5gaHCMwJYAJgBqGBZaaGEAvBHW0NkAPHJQJADmFBjEyQB8Kqo0aFDC+OkAwhBdUjBVACqZOWFoI2jE2WGTFnkFRTKtmsJZyzIAXHAT2wuCysrCWsQ2wEPaumMYAEaM0CDjS8cz2POLR-mFxa0ABQkYAwYAFd57A6QgCUiGUcDgDC0UGItz0XToDEBgLhJQ2qEw2DwhBIaEBCARiLgMBkY1SxX2CCEoU0EicLjk3HcKTS9TmEn2MCgRUUgjIVMRMkIaBAGCZkupyCG2BGTJZYUGw1GhymrUVAgl1JQKs+CuNiK2U32ILBEKOMgNVIEMJhygEpxm6ng50uYJupneVSpuuWfzWcE+c2O73DxSN0KOkdm3zgAGs0MkIMJEzk46V05ns7mw6tisogVTdFAAG7AZBofb6bqGJhiKqcbgkvhoKpa1U62Nl9ZNFqjtqtBNW5YAOQwIEbJeyyjhlMRyNR6PMVhs0EB1brDYoXOwcHxcBPaAAdMrtVfpwsANoPucLgC6MJUHrOFyuN3YrBoM8UAgBScC2uCFiQoITLgcQoKQZCsHynAxBaCADzdGKDz7A8EAQFYzRikILrwoit5yCILzmLW9bpGUpj3E8LzAvBdpQQ6n5UhuaJVMILz6saVTvIM-ikDQF7dIeaAlAg-EgTR0lCO4GzuO4cAABLdOQNIEOk9AQA8VjiMAYQwAQWC6ekB50SE8xWXABlgGE2RDBI8ByjAhBwNkRQ0hANIMvyxBUlU7jySA7Tuqcvp-miImsJA4mDr8w7Jl8Mapf86xgTZDYwWRcBqXAgxaMscAAO7pBZNbpBgcAAAbNliRjdMwrDjN2ZIUHhBGYMQrQNXAFjABmlXVRgtVUVAjVdaQADyYD+sQMjtSAQ0APxVlJdFNpirZte2nbErwZJ9qaIxvFlayTnZyTjq0+qkWuJorfAGDvGe6LvPuO1HpIGASFxiLFdSAB6W3kUMlFPEmDG6D9eU6RIDxAyoPGoWVFhfkAA) of the code. (Issue is on line 34)

I'm thinking it'd be best to have some extra helper types to assist with working with context, something like `StateMachineWithTypegen` along the lines of `MachineOptionsFrom<Machine, true>` (not sure how feasible that would be)

**Comment by Andarist** [comment, marked as answer]
This is a tricky one - and there a best answer for that: use instantiation expressions. However, this is the upcoming feature of TS and you can't quite use them today. Nevertheless... you can check out the demo here:
[TS playground](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220302#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs04BfOAMyghDgCIAPAZxizQYG4AoUSWIjgBXFmgCSxbFDDoYAGhFiAymgA2aZDGgVqtegwACrdtgD06fDG59w0eACo4GFnABKmLXrqNLWmzwwAJ5gZABiwOoAJgBqGGrCaK4AvO6a0FEAPGxQJADmihjEQQB8vHxSaFBU+GQAwhCVTDCZACoR0a5ozWjEUa4darHxiSwliDxw1JFDLABccO0z-bzkPDxUwsRawI1KaK0YAEZh0CBtg-1w3dh9A8txCUklABQkwDDA8Zfziz8AlBMpnJhFBiPtJNJZGgYC8XoDkiVJlMUJZsHhCCQ0C8EMiUXAYCxWiEkgsEJQXEIjMY2BwzMFQnlegwFjAoIkKPI8SiWIQ0CAMGTufjkI1sM0yRTXA0mi0lp0Slz8VNyErlaLZULlSiqMtfu9Pt89Wr8arueR-sj-qt1gzwmdVFAAG7AZBoNrjVIeGCg4jE0KZO0QKj7Q4nM4esrrboCeCbbafPaiNCXTLI+VDR6ja49O5wS5ZpJqjNRHO3K4AazQQWDfweIyScFSVZrIZLhZYPFeYmdrrQC1OUBAjpdbou9aeY0UuuiADkMCB+3XooDccCYb79qoNFpoC8e6O0Ip96Y0AjxrTsAA6DXimBXmezADaj6i88XAF1rTw1hstjs9mUOgUzOHFkQNL41B+LlyC1OAIKNToyWcBZiGEEAjiqLg4COBYjggCANCKChVlXZENTYagHSqQ8m1DY5BxAN5iA+SCAV4dcfTBOAXm5TIqDOJFtUyS4GjsUgpDgA8+2SBABKHEc+0oMwhJRTIzHkkBVO-X94wA8FRLoSAJLlAsGy6XMrjMydXjXKSaL7GC4Okt0B2o3sx3bcyowtIEUEaSiMEuOjk0ufcHLdRQGAwBhvymMwzHxAA9AB+ciAvgE5lhCsQwpco9GCOWKOLgEFuLQtQ1BtIA)

What can we do about this today though? Well... we could just expose `ResolvedTypegenFrom` type and allow it to be used like this:
[TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOASWOyjHVoDEoIQKBlGLXfI0i6rXppaFACoBPMGgDmaYgBFgAZwwAjADZoAJnAC+cAGYsQcAOQAPZV2xmA3AChQkWIjgBXZWkFo6DCp5o7GhayDDQ+kYm5gACVjZoAPTo+DD2TuDQ8ABUcBjKcABKmGFRrOYpYekOMFJocJLScsTMrMFQAG7AyGgAPGIAfHAAvA1waBbYxNoFPn4ivr0YxBIUy6t5K2tbcCSGvg11zUMA-IdN8nAAXHCkHb4ONXVwjMAh2gBqGBruaAWjxWQ0G0vWsUBIMm2EgGjicNF8hnw9QAwhB4ZN+q93gUJlMZi83hpPt9fsohkhDISZjcxFiiQU9I9DO5iGFgGiPF4xOpGNAQJiqTjJvJ8XTiT8-gMABQkYAwYDfMXKGlKgCUiAccDgDHcUGInO88PmMClUvVwyGqEw2DwhBIaClCE1WrgMGUjT+NwQBnybjMMTi1m4iVqF2IZhuMCgvzsrowMmV5gwAHc-qw0NyZGY4AAfJPECAwAi+TPZvRkZ1a5SENAgDBeysuoHomBen0FVEtgXYgaN8uN5vYSYNl0uynYm6y+WKwV9516VWqhyMp7SOA4CQe5ojc6yeStEDtLo9XrFGC64ge3qhtAQQwG7lqXlQEADXsOCYueDM1nyjmBMVemdWkqS+CUhTxAoxTA0kKy1ED3nGYVpgKABrNAJDvBpoJJP4d3QzD7wQokYL+BxpWdLxOm6NAbjmYQmBMXpOG4W0+D6TshxgbsSNwskoQEzYNnWQSNy3eQ3zgowqQAOQwEBaOwqkHHVJ0tR1PUDWCUJwigKUqOPcg4CDbARiGEy0FVRwVx-NkOXYdNnxAR1dmIOUFQ0JV9C9Vz3JnCc3HrW53BANRfFjNQbjUCAIC0ZZYwMBcNS1ZtrCMPkjxondAkfJyZTc6dPMFKznQ0-VekMPle1HXoxVRTJSBoYzfEM4YEEql9Mp6AxEiGRJEjgAAJXwjKLep6AgTRa12AoiywV1i2a6iek2XQxrgCawAKbQ0TMeA6xgQg4G0X5XQgV0jnkZ1ekSDrX2spkWTs-U6tYSBGu44jxVJJDIIJd5SLJFyDKyvQRzgfq4FRdwiTgVM4AIDB7jyOAAAN6IYXwD36Nj7QoaLYswYgBlRuANGAdC4fqRHkY6tHcdIAB5MA-2IZQD1Jk5KJami6KNBisaYlibV4e1ek43EeO+yVBJEoTZZ2MTLuJ3skrUlA0TSjAxWyrwxX0nmegoMwMDMEqtUhl0AD0uZSzX4CfKldbQfWQaN8w1DNxwyuCjQNGsoA)

But that is somewhat tedious as you have to fill all other "spots" with `any`s or something. I've tried to prototype a bunch of helpers for making this easier but everything seemed to be a little bit clunky.

A somewhat best solution I've figured out is something like this:
[TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgYygUwIY0wWWzQAtgA7TAGhTgEky8owtGAxKCEa1AZRlwJFSFbnQaYmLCaIAqATzCYA5pjIARYAGdsAIwA2mACbUAvnABmHEHABEAD0388NgNxJQkWIjgBXTZnpGZkwYSl9-HkwDNBhoODNLTlsAAQcnTAB6LCIYV3dwaHgAKjhsTTgAJRwYiytbbJi8pBgFTDgAQTI5Vh8yGOAIMjgAXjgACgA6KewoJU0ALlKugG0AXQBKEYA+Jbk3FsVKzE0IPQA3TAA1bD0fTAAeGR3RmThMOzwyQ3LO7t7+wZwAD8RxgPigZHkikeO0WMjczVaRxO5yMUOUqnYnBhI2Rpwu11uDyebw+qm+cHI5gklRoIIqpM+FMCEmCjHu2C6YU5cm5XN2YSpNJkVRRF0M6M0+BC2C2NFQIJFx3xaNaUpl8rgiwoFygNG1mF1+yRos0PjAejk6JUZHuNBkslFKolaul-AAEqdDNJ7U7US7FOr+Li-eLrZirI9Q6rA27sJ69N6oHLnnjUYS7jj3kzykKoLSFZVGeTyizJCEJPc83AAPqCsjU-M1uuUhs05st6sdls8uWoQtltmVlvNsIdsfjuBKsUx45xvv9g26-VwHUSBEHNqsYBRQwZ44hzBoaCGe6OKDkJR8uRbBHkRjmIhtADCgzwH0e2935WzJbgX8TfdNB2ZBUHMHdEwWKcAO+NwTCQJBzH+GABiGPxMBkXRWGgEBPwgilfy+coYKArYxnIYAUJuGCoJkGjNlAuAWHBND-EHFgxjGTZhgXVAMBwPBCBIchMDGRj+zgGBNElRYEDMMpEE1CSbGSVJHAEDJNxtGxFhgKA7jcCSJP4OZFhsbAAHdlRADDsCUGw4AAH1sTkIBgYgJEw+zDKMkxRAkzQSEwEBsFkpT+2PcQPlk+TylfKKYDw78tn8-s-PCyL3xgMKjP7cDv0WCiqL0GjUtQdKJJMdYaHWOCEM3OB8DkSIoDOYA0Aw1oksAm47h-MkiP-fDSMPTQzQtK1WhtO1UHYisoCxXC+AEIThAeeKsu6vdeuOFKBV2LYUpoTcIHMcJbJ0bCoBAJBbwQ95PHgJC+hQwF0JgmboOGnb+pzIbdyA2QYOLQaAGtMDkU6voBn7cXByGzro76iU0W6xn8Vr2swRYmpatqOvRLbSLCfLEwAOWwGy4RghiaGYiFzsiaJYigdGJHxqhxnUvBuJ2bnMAmTLswmUnvmWUWKZsjY6sQ5DULgHhOEwK6QDEmgiuAaj8M0UwctbSjNZK7XZNKbUfBAHR1zgHRFh0CBThwIYTDg2m+MGRwLBwvGsdxdDMMunDyLIA2te-Wq6ZCFjxk1e5zBw3j+3uGDXwKCgGDgDGOeGBA4+u72OrMDIE9QTOfc5QxJI8uBmAgMBykMQYAHJ4BCmASDgQw7kkiBJKm1QY4yXOQAXcP4Nll75eTzhIDTxKkZhlGQYpEifrIxjS463XFJL9msZx5rd4Jrr556lG7qq7f0Hd+BsGB0Z3vwtnMc3lybHD1AMgyCSAD0gRoSKPaXXwr7fwMEn4czCDYHQb8fL0yGGQHweg9B1SAA)

But I'm not super fond of it and it would become quite complex if we'd have to handle all the possible input types here. Also, note that it's currently somewhat simplified because only your context is generic - but if the event type would be generic too... that would become even harder to pull off.

You can also watch a video of me going through the whole reasoning process about this case here: https://twitter.com/AndaristRake/status/1506236097042472960

**Reply by SimeonC** [reply]
Thanks for the video, was really helpful and I learnt a lot about typescript infer which I kinda intuitively understood how it worked but hadn't figured out how to reason about it (and you got my name pronunciation correct!).

I did simplify this down quite a bit and I do use generics in the Events as well so I think for now I'll just live with sprinkling any everywhere and wait for TS 4.7.

**Reply by Andarist** [reply]
Could you share how you are working around this issue now? I will continue thinking about how the situation could be improved for cases like this.

**Reply by SimeonC** [reply]
Essentially I'm just doing this for now [TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOASWOyjHVoDEoIQKBlGLXfI0uAL5wAZixBwA5AA8Azl2wSA3AChQkWIjgBXGWmq16aGBR1p2aADZpkMaIJFjJAAVny0AenT4YS1eGjwAFRwGDJwAEqYNg6skl42vsowAJ5gaHCMwJYAJgBqGBZaaGEAvBHW0NkAPHJQJADmFBjEyQB8Kqo0aFDC+OkAwhBdUjBVACqZOWFoI2jE2WGTFnkFRTKtmsJZyzIAXHAT2wuCysrCWsQ2wEPaumMYAEaM0CDjS8cz2POLR-mFxa0ABQkYAwYAFd57A6QgCUiGUcDgDC0UGItz0XToDEBgLhJQ2qEw2DwhBIaEBCARiLgMBkY1SxX2CCEoU0EicLjk3HcKTS9TmEn2MCgRUUgjIVMRMkIaBAGCZkupyCG2BGTJZYUGw1G7z+a1aioEEupKBVnwVJsRWym+xBYIhRxkhqpAhhcNZzWSITC+m6hiYYiqnG4JL4aCqWtVOt+qwBTRa8baBoEpxm6ng50uYJupneVSphxyeuKcE+c2OutjMmN0KOpdm3zgAGs0MkIMJa0Wq3Ayi22x3C8ti06gVTdFAAG7AZBofa+rFGbrMVhBtyhskRs0jN4x-7rROJ1qtGvW5YAOQwIFnneWyjhlMRyNR6PMVhs0EB46nM4oXOwPY2P80AAOmVbVgNPBYAG1IIvK8AF0YRUFMzguK4bnYVg0GeKAQApOA7XBCxIUEJkCOIUEiMhMj5TgYgtBAB5ujFB59geCAICsZoxSEV14URMC5BEF5zEnad0jKUx7ieF5gQo+1iMdJCqSfNEqmEF4DRNKp3kGfxSBoOAv3EkoEA03DRO-NAhHcDZ3HcOAAAlunIGkCHSegIAeKxxGAMIYAILA3PSYyZxCeZgrgTywDCbIhgkeA5RgQg4GyIoaQgGkGX5YgqSqdxzJAdplBQzN0LRXTWEgAzRkHFY93rL4K13fV8NC9IBAtOB7LgQYtGWOAAHd0kCid0gwOAAAN539JdAzGddSAodjOMwYhWkmuALGAFshpGjAxuEqApsWtAAHkwGzYgZGXEBNoAfjHborLnTFZqgW7VxDXgN0jT4dy7Pdj3C5JD2Te8qUE+AMHeHt0XeT9nvEigJAwCRlMRHrqQAPUegShiEp460k3QEfalGHnRlRVLo-qLGQoA). It's not pretty or elegant but it does the job for now.

I have to get a bunch of "non-refactor" work done so that's all the time I have for this development cycle. Hopefully the next time I come around to this should be when ts 4.7 is out and I can use instantiation expressions which I think will be the best and cleanest solution.
